src/cargo/util/config.rs

```diff
@@ -701,14 +793,538 @@ impl Config {
+    }
+}
+
+/// Key for a configuration variable.
+#[derive(Debug, Clone)]
+struct ConfigKey(Vec<ConfigKeyPart>);
+
+impl ConfigKey {
+    fn from_str(key: &str) -> ConfigKey {
+        ConfigKey(
+            key.split('.')
+                .map(|p| ConfigKeyPart::Part(p.to_string()))
+                .collect(),
+        )
+    }
+
+    fn join(&self, next: ConfigKeyPart) -> ConfigKey {
+        let mut res = self.clone();
+        res.0.push(next);
+        res
+    }
+
+    fn to_env(&self) -> String {
+        format!(
+            "CARGO_{}",
+            self.0
+                .iter()
+                .map(|p| p.to_env())
+                .collect::<Vec<_>>()
+                .join("_")
+        )
+    }
+
+    fn to_config(&self) -> String {
+        self.0
+            .iter()
+            .map(|p| p.to_config())
+            .collect::<Vec<_>>()
+            .join(".")
+    }
+
+    fn last(&self) -> &ConfigKeyPart {
+        self.0.last().unwrap()
+    }
+}
+
+impl fmt::Display for ConfigKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.to_config().fmt(f)
+    }
+}
+
+/// Internal error for serde errors.
+#[derive(Debug)]
+pub struct ConfigError {
+    message: String,
+    definition: Option<Definition>,
+}
+
+impl ConfigError {
+    fn new(message: String, definition: Definition) -> ConfigError {
+        ConfigError {
+            message,
+            definition: Some(definition),
+        }
+    }
+
+    fn expected(key: &str, expected: &str, found: &ConfigValue) -> ConfigError {
+        ConfigError {
+            message: format!(
+                "`{}` expected {}, but found a {}",
+                key,
+                expected,
+                found.desc()
+            ),
+            definition: Some(Definition::Path(found.definition_path().to_path_buf())),
+        }
+    }
+
+    fn missing(key: String) -> ConfigError {
+        ConfigError {
+            message: format!("missing config key `{}`", key),
+            definition: None,
+        }
+    }
+
+    fn with_key_context(self, key: String, definition: Definition) -> ConfigError {
+        ConfigError {
+            message: format!("could not load config key `{}`: {}", key, self),
+            definition: Some(definition),
+        }
+    }
+}
+
+impl std::error::Error for ConfigError {
+    fn description(&self) -> &str {
+        self.message.as_str()
+    }
+}
+
+impl fmt::Display for ConfigError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if let Some(ref definition) = self.definition {
+            write!(f, "error in {}: {}", definition, self.message)
+        } else {
+            self.message.fmt(f)
+        }
+    }
+}
+
+impl de::Error for ConfigError {
+    fn custom<T: fmt::Display>(msg: T) -> Self {
+        ConfigError {
+            message: msg.to_string(),
+            definition: None,
+        }
+    }
+}
+
+// TODO: AFAIK, you cannot override Fail::cause (due to specialization), so we
```

Comment:
I'm a little confused by this, couldn't the `failure::Error` be stored directly in `ConfigError`?

Reply:
It could, but it is not possible to implement `Fail::cause()` (because there is a blanket impl of `Fail` for `std::Error` you get a conflicting implementation error).  I could maybe write a method that creates a `failure::Error` by chaining `ConfigError` with the inner Fail error, but it seems like it would be easy to accidentally forget to call that method in the right places (`?` or `into` would circumvent it causing the chain to be lost).

Reply:
Ah ok, I think I see now what's going on. Perhaps the raw failure object could be stored for now though and the rendering could happen in the `Display` impl? Other than that though I think this is functionally fine in that there's not too many underlying errors coming out here anyway